test/http_server_test2/http_server_test2.cpp

```diff
@@ -0,0 +1,363 @@
+                IConnectionTimeoutModifier& timeout_modifier,
+                const std::string& url,
+                bool first_part,
+                bool last_part,
+                const std::unordered_map<std::string, std::string>& headers,
+                const std::unordered_map<std::string, std::string>& request_parameters,
+                const std::vector<uint8_t>& content,
+                MIMEParser& mime) {
+            (void) timeout_modifier;
+            (void) url;
+            (void) first_part;
+            (void) headers;
+            (void) content;
+            (void) mime;
+
+            if (last_part)
+            {
+                std::size_t size = 0;
+
+                try
+                {
+                    size = std::stoul(request_parameters.at("size"));
+                }
+                catch (...)
+                {
+                }
+
+                if (size == 0)
+                {
+                    response.reply(std::make_unique<responses::StringResponse>(ResponseCode::Expectation_Failed,
+                                                                               "Request parameter 'size' must be > 0"),
+                                   false);
+                }
+                else
+                {
+                    response.reply(std::make_unique<SendBlob>(size), false);
+                }
+            }
+        };
+
+        auto upload = [web_root, uploads](
+                IServerResponse& response,
+                IConnectionTimeoutModifier& timeout_modifier,
+                const std::string& url,
+                bool first_part,
+                bool last_part,
+                const std::unordered_map<std::string, std::string>& headers,
+                const std::unordered_map<std::string, std::string>& request_parameters,
+                const std::vector<uint8_t>& content,
+                MIMEParser& mime) {
+            (void) timeout_modifier;
+            (void) headers;
+            (void) request_parameters;
+            (void) url;
+
+            if (first_part)
+            {
+                // Prepare mime parser to receive data
+                mime.detect_mode(headers.at(CONTENT_TYPE), std::stoul(headers.at(CONTENT_LENGTH)));
+            }
+
+            auto form_data = [uploads, &last_part, &response](const std::string& name,
+                                                        const std::string& actual_file_name,
+                                                        const MIMEParser::BoundaryIterator& begin,
+                                                        const MIMEParser::BoundaryIterator& end,
+                                                        const bool file_start,
+                                                        const bool file_close) {
+                // Store the file in web_root/uploads
+                static Path path{};
+                static std::ofstream to_save;
```

Comment:
These can't be static, I think - same reasoning as in MIMEParser re. multithreading. Since this is a lambda, they have to be passed in by reference so they can be reused over several calls (and must thus be kept alive in the outside scope).

Reply:
I am not sure if these can't remain static. If I would launch another http-server, there would be other lambda functions defined for managing the uploads, since this is not a function inside the smooth library but a 'potential' user code.
If I would have to pass these in by reference, I think it would be a messy thing passing it from `App` to the lambda function `upload` and from there to the lambda function `form_data`. 
Or am I missing something here?

Reply:
Have a look at http://cpp.sh/9j66l and run it. You'll see that it prints "1", "2" even though the example is making copies of the lambda into `b` and `c`. So, if you have statics inside your lambda it'll be shared between copies, just like a class.

The example reuses the lambdas for both the secure and insecure server, contrary to you point that there will be different lambdas so the lambdas must be thread-safe.

I agree that having to pass additional arguments would be a bit messy. Would it perhaps be better to add a template argument to the HTTPServer class that specifies a type of which an object can be retrieved in the lambda...something like a session object?

Reply:
yes, you are right. I need to think again about how to solve it.

Reply:
An easy solution could be to pass a void* to the HTTPServer.on() method. But I guess this is more C-style rather than C++, right?